docs/provisioning.rst

```diff
@@ -0,0 +1,355 @@
+
+.. code:: python
+
+    def _get_application_id(self, ignore_final_states: bool = False) -> str:
+
+        if not self.application_id:
+            app = self._query_app_by_name(self.kernel_id)
+            state_condition = True
+            if type(app) is dict:
+                state = app.get('state')
+                self.last_known_state = state
+
+                if ignore_final_states:
+                    state_condition = state not in YarnProvisioner.final_states
+
+                if len(app.get('id', '')) > 0 and state_condition:
+                    self.application_id = app['id']
+                    self.log.info(f"ApplicationID: '{app['id']}' assigned for "
+                                  f"KernelID: '{self.kernel_id}', state: {state}.")
+            if not self.application_id:
+                self.log.debug(f"ApplicationID not yet assigned for KernelID: "
+                               f"'{self.kernel_id}' - retrying...")
+        return self.application_id
+
+
+    def get_shutdown_wait_time(self, recommended: Optional[float] = 5.0) -> float:
+
+        if recommended < yarn_shutdown_wait_time:
+            recommended = yarn_shutdown_wait_time
+            self.log.debug(f"{type(self).__name__} shutdown wait time adjusted to "
+                           f"{recommended} seconds.")
+
+        return recommended
+
+Registering your custom provisioner
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Once your custom provisioner has been authored, it needs to be exposed
+as an
+`entry point <https://packaging.python.org/specifications/entry-points/>`_.
+To do this add the following to your ``setup.py`` (or equivalent) in its
+``entry_points`` stanza using the group name
+``jupyter_client.kernel_provisioners``:
+
+::
+
+            'jupyter_client.kernel_provisioners': [
+                'rbac-provisioner = acme.rbac.provisioner:RBACProvisioner',
+            ],
+
+where:
+
+-  ``rbac-provisioner`` is the *name* of your provisioner and what will
+   be referenced within the ``kernel.json`` file
+-  ``acme.rbac.provisioner`` identifies the provisioner module name, and
+-  ``RBACProvisioner`` is custom provisioner object name
+   (implementation) that (directly or indirectly) derives from
+   ``KernelProvisionerBase``
+
+Deploying your custom provisioner
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The final step in getting your custom provisioner deployed is to add a
+``kernel_provisioner`` stanza to the appropriate ``kernel.json`` files.
+This can be accomplished manually or programmatically (in which some
+tooling is implemented to create the appropriate ``kernel.json`` file).
+In either case, the end result is the same - a ``kernel.json`` file with
+the appropriate stanza within ``metadata``. The *vision* is that kernel
+provisioner packages will include an application that creates kernel
+specifications (i.e., ``kernel.json`` et. al.) pertaining to that
+provisioner.
+
```

Comment:
Well, I can see a vision where a single `kernel.json` and some `metadata` (the default values), a `config_schema`. The consumer application would be free to use the schema and pass some `user_config` that  would override the default ones. How those `user_config` would be passed is prolly still to be discussed.

In that sense, I wonder if the additional stanza in the `kernel.json`should be:

- `config`-> `default_config`
- `config_schema`

The `user_config` should be not persisted in the kernel.json but could perfectly find its place in e.g. the metadata of the ipynb notebook.

Reply:
Again, this falls into the parameterization discussion.  However, I believe we'll need to state that any _required properties_ specified in the schema must have a default value since we can't assume that every consumer of a kernelspec is going to be able to provide parameters.

I agree that persisting user-provided values in the notebook itself is reasonable.  We'll likely want some association/namespacing of those attributes to the associated schema as well.  Then, when that notebook is executed with the same provisioner/schema, _those_ values could be pre-filled/used.  Still lots to work out on that front.

Reply:
Hmm - in thinking about this just now, I think we also need to separate _configuration_ from _parameters_.  With configuration, I view this as more static in nature.  It can be specified in the `kernel.json`, a config file, or something else - but that aspect of things is completely provisioner-specific and the provisioner is responsible for distinguishing configuration from parameters.  

Examples of configuration might be the resource manager hostname for _this_ installation.  In that case, you don't want to expose that to the end-user, but instead, it's a value that is required by every invocation of that particular provisioned kernel.

Reply:
Interested discussion on nomenclature. Let's say

- `resource_manager_hostname` has to be given to the provisioner, but is enforced by the kernelspec author and can not changed by the user
- `memory` has to be given to provisioner, has a default value of `1G` defined by the kernel author, but can be changed by the user with value `1G`, `5G` or `10G` as stated in a json schema defined by the kernel author.

I see them all as the same thing, by some can not be changed by the user.

Thinking loud, I could see name like `system parameter`/ `user parameter`  or `system config` / `user config`.

Alternative to system vs user could be static vs dynamic.

Reply:
I agree.  All parameters, _system_ or _user_, _static_ or _dynamic_, should be available to everyone.  Where that presentation takes place is, I believe, where the differences lie.  

I view system parameters as static values where you would not want them changed by the user and those would be oriented to the provisioner's underlying configuration system (which may or may not be trailtets-based). In some (typically single-user) cases, _everyone_ can be split into different roles, namely _admin_ and _user_, where a given _user_ is not exposed to a _system_ parameter.  In other cases, you might want various system parameters exposed to the user for modification.  Since any user-exposed parameters would need to be reflected in the schema and since it could be the case that _system parameters_ should also be available to users for modification, perhaps a _meta-propery_ like `is_hidden` is warranted.  This all implies, then, that the schema itself would need to be configurable (which I suppose is already the case in order to set appropriate enumerated types, etc.). 

Note, there's nothing preventing a provisioner from implementing its user-modifiable parameters based on an underlying configuration system (e.g., configurable traits) so long as the _runtime_ values override the configured values.

These are all the kinds of things that will require hashing out during _parameterized kernels_ discussions.

Reply:
> These are all the kinds of things that will require hashing out during parameterized kernels discussions.

Agree with the previous comment. fyi I will start experimenting ui in jlab and push more stanza just to make sense of the parametrisation options. I hope to come with a demo at the next server meeting.